### PR TITLE
Recreate state for rpcs

### PR DESCRIPTION
**Pull request by magicxyyz**
Support recreating archive state for RPCs like eth_call, eth_getBalance, etc.

### CODE REVIEW

File: arbitrum/recreatestate.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+				return nil, lastHeader, ErrDepthLimitExceeded
+			}
+		}
+		if logFunc != nil {
+			logFunc(header, currentHeader, false)
+		}
+		if currentHeader.Number.Uint64() <= genesis {
+			return nil, lastHeader, errors.Wrap(err, fmt.Sprintf("moved beyond genesis looking for state %d, genesis %d", header.Number.Uint64(), genesis))
+		}
+		currentHeader = bc.GetHeader(currentHeader.ParentHash, currentHeader.Number.Uint64()-1)
```
[the remark this answers is not in the document]
Comment (magicxyyz) [reply arbitrum/recreatestate.go:52]: fixed, thanks

File: arbitrum/recreatestate.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		stateDb, err = stateFor(currentHeader)
+		if err == nil {
+			break
+		}
+	}
+	return stateDb, currentHeader, nil
+}
+
+func RecreateBlock(ctx context.Context, bc *core.BlockChain, header *types.Header, stateDb *state.StateDB, blockToRecreate uint64, prevBlockHash common.Hash, logFunc StateBuildingLogFunction) (*types.Block, error) {
```
[the remark this answers is not in the document]
Comment (magicxyyz) [reply arbitrum/recreatestate.go]: yeah, the naming I used was terrible, hopefully it's better now - please let me know what you think, thanks!

**Review by PlasmaPower — DISMISSED** [review]

File: arbitrum/config.go [hunk 3]
```diff
@@ -67,6 +68,7 @@ var DefaultConfig = Config{
 	FilterTimeout:           5 * time.Minute,
 	FeeHistoryMaxBlockCount: 1024,
 	ClassicRedirect:         "",
+	MaxRecreateStateDepth:   0, // TODO
```
[the remark this answers is not in the document]
Comment (magicxyyz) [reply arbitrum/config.go]: changed and added special values:
 * -2 for uninitialized value that should be overwritten with corresponding default
*  -1 for infinite
* and 0 disables state recreation
Thanks for the suggestion :)

**Review by PlasmaPower — APPROVED** [review]
LGTM

**Pull request merged**